File: java/rocksjni/sim_cache.cc
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).
//
// This file implements the "bridge" between Java and C++ for
// ROCKSDB_NAMESPACE::SimCache.

#include "include/rocksdb/utilities/sim_cache.h"

#include <jni.h>

#include "include/org_rocksdb_SimCache.h"
#include "rocksjni/cplusplus_to_java_convert.h"

/*
 * Class:     org_rocksdb_SimCache
 * Method:    newSimCache
 * Signature: (JJI)J
 */
jlong Java_org_rocksdb_SimCache_newSimCache(JNIEnv* /*env*/, jclass /*jcls*/,
                                            jlong jhandle,
                                            jlong jcapacity,
                                            jint jnum_shard_bits) {
  auto* sptr_sim_cache = new std::shared_ptr<ROCKSDB_NAMESPACE::Cache>(
      ROCKSDB_NAMESPACE::NewSimCache(
          *reinterpret_cast<std::shared_ptr<ROCKSDB_NAMESPACE::Cache>*>(jhandle),
          static_cast<size_t>(jcapacity), 
          static_cast<int>(jnum_shard_bits)));
  return GET_CPLUSPLUS_POINTER(sptr_sim_cache);
}

/*
 * Class:     org_rocksdb_SimCache
 * Method:    disposeInternal
 * Signature: (J)V
 */
void Java_org_rocksdb_SimCache_disposeInternal(JNIEnv* /*env*/,
                                               jobject /*jobj*/,
                                               jlong jhandle) {
  auto* sptr_sim_cache =
      reinterpret_cast<std::shared_ptr<ROCKSDB_NAMESPACE::Cache>*>(jhandle);
  delete sptr_sim_cache;  // delete std::shared_ptr
}

File: java/src/main/java/org/rocksdb/SimCache.java
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).

package org.rocksdb;

/**
 * Simulation Cache
 */
public class SimCache extends Cache {

  /**
   * Create a new key only cache with a fixed size capacity. The cache is sharded
   * to 2^numShardBits shards, by hash of the key. The total capacity
   * is divided and evenly assigned to each shard.
   * numShardBits = -1 means it is automatically determined: every shard
   * will be at least 512KB and number of shard bits will not exceed 6.
   *
   * @param handle The native handle of the real cache
   * @param capacity The fixed size capacity of the key only cache
   * @param numShardBits The key only cache is sharded to 2^numShardBits shards,
   *     by hash of the key
   */
  public SimCache(final long handle, final long capacity, final int numShardBits) {
    super(newSimCache(handle, capacity, numShardBits));
  }

  private static native long newSimCache(final long handle, final long capacity, final int numShardBits);
  @Override protected final native void disposeInternal(final long handle);
}
